### PR TITLE
[RNMobile] Fix error when pasting deeply nested structure content

### LINKED ISSUE #18750: JavascriptException: RangeError: TaskQueue: Error with task : Maximum call stack size exceeded (native stack depth)
Sentry Issue: [JETPACK-ANDROID-7XY](https://a8c.sentry.io/issues/4262949664/?referrer=github_integration)

**Symbolicated stack trace:**
```
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
allocateTag - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1988
unmountHostComponents - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:5910
commitMutationEffects - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:6013
commitRootImpl - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:7060
completeUnitOfWork - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:7008
markRootSuspended$1 - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:6647
invalidateContextProvider - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:2121
scheduleUpdateOnFiber - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:6335
clz32Fallback - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-prod.js:1787
b.then$argument_1 - gutenberg-mobile/gutenberg/node_modules/react/cjs/react.production.min.js:18
_nestedChildLists.forEach$argument_0 - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Lists/VirtualizedList.js:1704
<global> - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Interaction/Batchinator.js:13
tasks.forEach$argument_0 - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Interaction/TaskQueue.js:67
<global> - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Interaction/InteractionManager.js:154
setInterval - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Core/Timers/JSTimers.js:230
_callTimer - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Core/Timers/JSTimers.js:92
_callTimer - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Core/Timers/JSTimers.js:116
callIdleCallbacks - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/Core/Timers/JSTimers.js:380
enqueueNativeCall - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:331
__guard$argument_0 - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:114
enqueueNativeCall - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:316
constructor - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:85
constructor - gutenberg-mobile/gutenberg/node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:61
```

```
JavascriptException: RangeError: TaskQueue: Error with task : Maximum call stack size exceeded (native stack depth)

This error is located at:
    in l
    in RCTView
    in Unknown
    in RCTScrollView
    in C
    in ScrollView
    in Unknown
    in u
    in S
    in AnimatedComponent(S)
    in Unknown
    in RCTView
    in Unknown
    in E
    in v
    in RCTView
    in Unknown
    in AnimatedComponent(View)
    in Unknown
    in c
    in AnimatedComponent(c)
    in Unknown
    in Unknown
    in f
    in Unknown
    in RCTView
    in Unknown
    in Unknown
    in B
    in RCTView
    in Unknown
    in RCTView
    in Unknown
    in Unknown
    in RCTView
    in Unknown
    in Unknown
    in R
    in WithPreferredColorScheme(R)
    in Unknown
    in WithSelect(WithPreferredColorScheme(R))
    in RNCSafeAreaProvider
    in Unknown
    in Unknown
    in Unknown
    in Unknown
    in WithRegistryProvider(Component)
    in Unknown
    in Unknown
    in Unknown
    in Unknown
    in Unknown
    in WithRegistryProvider(Component)
    in h
    in H
    in WithDispatch(H)
    in Unknown
    in WithSelect(WithDispatch(H))
    in y
    in RNGestureHandlerRootView
    in Unknown
    in F
    in WithDispatch(F)
    in Unknown
    in WithSelect(WithDispatch(F))
    in k
    in RCTView
    in Unknown
    in RCTView
    in Unknown
    in E, js engine: hermes, stack:
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
Gn@1:172289
ba@1:200191
ka@1:201232
Si@1:209533
yi@1:209149
oi@1:206239
xt@1:173349
ni@1:203510
enqueueSetState@1:218168
anonymous@1:145593
anonymous@1:429493
anonymous@1:440377
value@1:381368
S@1:380068
anonymous@1:249595
w@1:248589
N@1:248752
callReactNativeMicrotasks@1:250665
value@1:234104
anonymous@1:233234
value@1:233981
value@1:233145
value@1:233001

    at com.facebook.react.modules.core.ExceptionsManagerModule.reportException(ExceptionsManagerModule.java:72)
    at java.lang.reflect.Method.invoke(Method.java)
    at com.facebook.react.bridge.JavaMethodWrapper.invoke(JavaMethodWrapper.java:372)
    at com.facebook.react.bridge.JavaModuleWrapper.invoke(JavaModuleWrapper.java:188)
    at com.facebook.react.bridge.queue.NativeRunnable.run(NativeRunnable.java)
...
(6 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by fluiddot**
## What?

Fixes an error related to pasting content with complex structures that leads to hitting the maximum call stack size specified in Hermes ([reference](https://github.com/wordpress-mobile/WordPress-Android/issues/18750#issuecomment-1776123223)).

## Why?

Fixes https://github.com/wordpress-mobile/WordPress-Android/issues/18750 and https://github.com/wordpress-mobile/gutenberg-mobile/issues/6123.

## How?

The issue was produced due to the fact that the implementation of `Node.contains` was recursive. This function is used as part of the pasting functionality, so when a deeply nested structure is processed, the call stack could grow until it reaches the limit and crashes the editor.

https://github.com/WordPress/gutenberg/blob/ae769e92bf388c3abdd75fe583b74bc1f4c30163/packages/react-native-editor/src/jsdom-patches.js#L42
https://github.com/WordPress/gutenberg/blob/1501effca4cf99fd39c30b6f6d0f443e6b090c55/packages/blocks/src/api/raw-handling/utils.js#L128-L134
https://github.com/WordPress/gutenberg/blob/1501effca4cf99fd39c30b6f6d0f443e6b090c55/packages/blocks/src/api/raw-handling/utils.js#L153-L158
https://github.com/WordPress/gutenberg/blob/1501effca4cf99fd39c30b6f6d0f443e6b090c55/packages/blocks/src/api/raw-handling/paste-handler.js#L212
https://github.com/WordPress/gutenberg/blob/1501effca4cf99fd39c30b6f6d0f443e6b090c55/packages/block-editor/src/components/rich-text/index.native.js#L452-L453

The solution proposed is to replace the implementation with a non-recursive approach. In fact, the implementation has been copied from the original polyfill we use in WordPress core for the function `Node.contains` ([reference 1](https://core.trac.wordpress.org/browser/trunk/package.json#L166), [reference 2](https://github.com/JakeChampion/polyfill-library/blob/master/polyfills/Node/prototype/contains/polyfill.js#L3-L16), [reference 3](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/script-loader.php#L99)). This polyfill is no longer used but we keep a reference ([reference](https://core.trac.wordpress.org/changeset/56646)).

**⚠️ NOTE:** As you can see above, the function `deepFilterNodeList` is also recursive, hence we might still get the error after applying this fix. However, it's likely that would happen in very edge cases.

## Testing Instructions

1. Create/open a post in the app.
2. Copy the following HTML code with a deeply nested structure.
<details><summary>Click here to display the HTML code</summary>

    <ul>
        <li>1
            <ul>
                <li>2
                    <ul>
                        <li>3<ul>
                                <li>4<ul>
                                        <li>5<ul>
                                                <li>6<ul>
                                                        <li>7<ul>
                                                                <li>8<ul>
                                                                        <li>9<ul>
                                                                                <li>10<ul>
                                                                                        <li>11<ul>
                                                                                                <li>12<ul>
                                                                                                        <li>13<ul>
                                                                                                                <li>14
                                                                                                                    <ul>
                                                                                                                        <li>15
                                                                                                                            <ul>
                                                                                                                                <li>16
                                                                                                                                    <ul>
                                                                                                                                        <li>17
                                                                                                                                            <ul>
                                                                                                                                                <li>18
                                                                                                                                                    <ul>
                                                                                                                                                        <li>19
                                                                                                                                                            <ul>
                                                                                                                                                                <li>20
                                                                                                                                                                    <ul>
                                                                                                                                                                        <li>21
                                                                                                                                                                            <ul>
                                                                                                                                                                                <li>22
                                                                                                                                                                                </li>
                                                                                                                                                                            </ul>
                                                                                                                                                                        </li>
                                                                                                                                                                    </ul>
                                                                                                                                                                </li>
                                                                                                                                                            </ul>
                                                                                                                                                        </li>
                                                                                                                                                    </ul>
                                                                                                                                                </li>
                                                                                                                                            </ul>
                                                                                                                                        </li>
                                                                                                                                    </ul>
                                                                                                                                </li>
                                                                                                                            </ul>
                                                                                                                        </li>
                                                                                                                    </ul>
                                                                                                                </li>
                                                                                                            </ul>
                                                                                                        </li>
                                                                                                    </ul>
                                                                                                </li>
                                                                                            </ul>
                                                                                        </li>
                                                                                    </ul>
                                                                                </li>
                                                                            </ul>
                                                                        </li>
                                                                    </ul>
                                                                </li>
                                                            </ul>
                                                        </li>
                                                    </ul>
                                                </li>
                                            </ul>
                                        </li>
                                    </ul>
                                </li>
                            </ul>
                        </li>
                    </ul>
                </li>
            </ul>
        </li>
    </ul>

</details>

3. Paste the content in a Paragraph block.
4. Observe that a List block is created and that the editor doesn't throw an error.
5. Observe that the List block doesn't render all its items due to [this limitation](https://github.com/WordPress/gutenberg/pull/54382).

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 
N/A

### CODE REVIEW

File: test/native/jest.config.js [hunk 1]
```diff
@@ -18,6 +18,19 @@ const transpiledPackageNames = glob( 'packages/*/src/index.{js,ts}' ).map(
...
+// Reference: https://github.com/WordPress/gutenberg/issues/55652
+const RAW_HANDLING_UNSUPPORTED_UNIT_TESTS = [
+	'html-formatting-remover',
+	'phrasing-content-reducer',
+	'ms-list-converter',
+	'figure-content-reducer',
+	'special-comment-converter',
+	'normalise-blocks',
+	'image-corrector',
+];
```
Comment (fluiddot) [inline test/native/jest.config.js:32]: We'll enable these unit tests when addressing https://github.com/WordPress/gutenberg/issues/55652.

**Review by geriux — APPROVED** [review]
LGTM! I've tested this on both iOS and Android, I also ran the unit tests for the `raw-handling` functionality. 🚀

**Comment by geriux** [comment]
❓ Quick question, with the test HTML provided in the PR description, should the option to "Ungroup block" work? If I tap it, it doesn't do anything 😅 

<kbd></kbd>

**Comment by fluiddot** [comment]
> ❓ Quick question, with the test HTML provided in the PR description, should the option to "Ungroup block" work? If I tap it, it doesn't do anything 😅

Yeah, it should ungroup the block. I'll take a look 👀, thanks for sharing it !

**Comment by fluiddot** [comment]
> > ❓ Quick question, with the test HTML provided in the PR description, should the option to "Ungroup block" work? If I tap it, it doesn't do anything 😅
> 
> Yeah, it should ungroup the block. I'll take a look 👀, thanks for sharing it !

I managed to reproduce it. Looks like there's a bug related to the ungrouping logic used in the `WarningMaxDepthExceeded`. It assumes that the block can be replaced with its inner blocks, but this is not the case for some of the blocks like the List block. I'll open a GitHub issue and work on a solution. In the meantime, we could merge this PR as-is because both issues are unrelated.

**Comment by geriux** [comment]
> I managed to reproduce it. Looks like there's a bug related to the ungrouping logic used in the `WarningMaxDepthExceeded`. It assumes that the block can be replaced with its inner blocks, but this is not the case for some of the blocks like the List block. I'll open a GitHub issue and work on a solution. In the meantime, we could merge this PR as-is because both issues are unrelated.

That sounds good to me, thank you for investigating it!

**Comment by fluiddot** [comment]
> [...] I'll open a GitHub issue and work on a solution. In the meantime, we could merge this PR as-is because both issues are unrelated.

For reference: https://github.com/WordPress/gutenberg/issues/55738

**Pull request merged**